### PR TITLE
peak valley generation from rip data.

### PR DESCRIPTION
**Pull request by kauser-cse-buet**
Implemented peak valley generation from rip data with documentation. Unit test will be added later.

### CODE REVIEW

File: .idea/CerebralCortex.iml [hunk 1]
```diff
@@ -2,15 +2,14 @@
 <module type="JAVA_MODULE" version="4">
   <component name="FacetManager">
     <facet type="Python" name="Python">
-      <configuration sdkName="Python 3.5.2 virtualenv at ~/.virtualenvs/CerebralCortex" />
+      <configuration sdkName="" />
```
Comment (twhnat) [inline .idea/CerebralCortex.iml]: This file should not be modified and committed

File: cerebralcortex/data_processor/cStress.py [hunk 2]
```diff
@@ -91,11 +92,33 @@ def cStress(raw_ecg: DataStream,
     print('mag-deviations-length:', len(accelerometer_win_mag_deviations.datapoints))
     print('accel_activity-length:', len(accel_activity.datapoints))
 
+    # rip features
+    smoothing_factor = 5
```
Comment (twhnat) [inline cerebralcortex/data_processor/cStress.py]: These values and parameter should be made into default named parameters of the function `compute_peak_valley`

File: cerebralcortex/data_processor/cStress.py [hunk 3]
```diff
@@ -91,11 +92,33 @@ def cStress(raw_ecg: DataStream,
...
+                                                                         fs = rip_sampling_frequency,
+                                                                         smoothing_factor = smoothing_factor,
+                                                                         time_window = time_window,
+                                                                         expiration_amplitude_threshold_perc = expiration_amplitude_threshold_perc,
+                                                                         threshold_expiration_duration=threshold_expiration_duration,
+                                                                         inspiration_amplitude_threshold_perc = inspiration_amplitude_threshold_perc,
+                                                                         max_amplitude_change_peak_correction = max_amplitude_change_peak_correction,
+                                                                         min_neg_slope_count_peak_correction = min_neg_slope_count_peak_correction)
+
+    # r-peak datastream computation
```
Comment (twhnat) [inline cerebralcortex/data_processor/cStress.py]: Duplicated code

File: cerebralcortex/data_processor/signalprocessing/rip.py [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+    :param rip:
+    :param fs:
+    :param time_window:
+    :param expiration_amplitude_threshold_perc:
+    :param threshold_expiration_duration:
+    :param max_amplitude_change_peak_correction:
+    :param min_neg_slope_count_peak_correction:
+    """
+    data = rip.datapoints
+    sample = np.array([i.sample for i in data])
```
Comment (twhnat) [inline cerebralcortex/data_processor/signalprocessing/rip.py]: You should be thinking in terms of `List[DataPoint]` and not breaking up the logic based on how Matlab handles everything.

File: cerebralcortex/data_processor/signalprocessing/rip.py [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+    timestamp = np.array([i.start_time for i in data])
+
+    sample = smooth(sample, smoothing_factor) # Moving average filter with default smoothing factor = 5.
```
Comment (twhnat) [inline cerebralcortex/data_processor/signalprocessing/rip.py]: Sample should be of type `List[DataPoint]`

File: cerebralcortex/data_processor/signalprocessing/rip.py [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+    window_length = round(time_window * fs)
+    mac = get_moving_average_curve(sample, window_length=window_length)
```
Comment (twhnat) [inline cerebralcortex/data_processor/signalprocessing/rip.py]: Should return type List[DataPoint]

File: cerebralcortex/data_processor/signalprocessing/rip.py [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+
+    up_intercept, up_intercept_indices, down_intercept, down_intercept_indices = get_intercepts(sample=sample, mac=mac, window_length=window_length)
```
Comment (twhnat) [inline cerebralcortex/data_processor/signalprocessing/rip.py]: Indices are not necessary to return unless you have a good reason for it.  Thoughts?

File: cerebralcortex/data_processor/signalprocessing/rip.py [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+
+    for i in valley_indices:
+        valley_datapoints.append(DataPoint.from_tuple(sample[i], timestamp[i]))
+
+    peak_datastream = DataStream.from_datastream([rip])
+    peak_datastream.datapoints = peak_datapoints
+    valley_datastream = DataStream.from_datastream([rip])
+    valley_datastream.datapoints = valley_datapoints
+
+    # TODO: graph plot
```
Comment (twhnat) [inline cerebralcortex/data_processor/signalprocessing/rip.py]: Unused code should be removed

File: cerebralcortex/data_processor/signalprocessing/rip.py [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+                                  valley_indices: np.ndarray) -> [np.ndarray, np.ndarray]:
+    """
+    Filter out too close valley peak pair.
+
+    :return peak_indices_updated, valley_indices_updated:
+    :param peak_indices
+    :param valley_indices
+    """
+    peak_to_valley_index_list = [(peak_indices[i] - valley_indices[i]) for i in range(len(peak_indices))]
+    valid_peak_to_valley_index_list = [ i for i in range(len(peak_to_valley_index_list)) if peak_to_valley_index_list[i] >= 6] # minimum peak to valley distance assume valid if > 0.3 sec. floor(64/3*0.31 sec) = 6 points.
```
Comment (twhnat) [inline cerebralcortex/data_processor/signalprocessing/rip.py]: This type of filtering is not obvious.  All the logic should be converted over to code that operates on DataPoints and thus can use the appropriate temporal notions.  This is also not general if someone else wants to utilize these algorithms.

File: cerebralcortex/data_processor/signalprocessing/rip.py [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+        values = sample[down_intercept_indices[i]: up_intercept_indices[i]]
+        valley = min(values)
+        valleys.append(valley)
+        valleyIndex = down_intercept_indices[i] + list(values).index(valley)
+        valley_indices.append(valleyIndex)
+
+    return peaks, peak_indices, valleys, valley_indices
+
+
+def filter_intercept_outlier(up_intercept_indices: np.ndarray,
```
Comment (twhnat) [inline cerebralcortex/data_processor/signalprocessing/rip.py]: This whole function is difficult to understand.  I do not see the need for your up and down symbols

File: cerebralcortex/data_processor/signalprocessing/rip.py [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+
+        if dictInterceptIndices[keys[i]] == upSymbol:
+            uii.append(keys[i])
+        elif dictInterceptIndices[keys[i]] == downSymbol:
+            dii.append(keys[i])
+
+        i += 1
+
+    i = 0
+    while uii[i] < dii[0]: # Targetted to start from down intercept. That's why trim up intercepts out where up intercept is less than first down intercept.
```
Comment (twhnat) [inline cerebralcortex/data_processor/signalprocessing/rip.py]: Do not do this

File: cerebralcortex/data_processor/signalprocessing/rip.py [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+    Moving Average Centerline curve intersects breath cycle twice. Once in the inhalation branch (Up intercept) and in the exhalation branch (Down intercept).
+
+    :return up_intercept, up_intercept_indices, down_intercept, down_intercept_indices:
+    :param sample:
+    :param mac:
+    :param window_length:
+    """
+    up_intercept, up_intercept_indices, down_intercept, down_intercept_indices = [], [], [], []
+
+    for i in range(1, len(mac)):
```
Comment (twhnat) [inline cerebralcortex/data_processor/signalprocessing/rip.py]: Improper array indexing by going from 1 to n and the last point will likely be fogotton

**Review by twhnat — CHANGES_REQUESTED** [review]
This is a great start for a first-pass at the algorithm.  I would like you to rethink the implementation to base it on lists of DataPoints instead of np.arrays.  The justification is to alway maintain the timestamp-data pairing unless they need split apart within a single function.  Lets chat about your questions before moving forward with the improvements.

File: .idea/CerebralCortex.iml (hunk 1 above)
Comment (kauser-cse-buet) [reply .idea/CerebralCortex.iml]: I committed these files accidentally. I push another commit with these files having their previous state.

**Pull request merged**